File: fixtures/scale/kotlin/large/src/cryptobench/util/Rng.kt
package cryptobench.util

/** xorshift64* generator; deterministic so runs are reproducible. */
class Rng(seed: Long) {
    private var state: Long = if (seed == 0L) 0x9E3779B97F4A7C15UL.toLong() else seed

    fun nextLong(): Long {
        var x = state
        x = x xor (x ushr 12)
        x = x xor (x shl 25)
        x = x xor (x ushr 27)
        state = x
        return x * 0x2545F4914F6CDD1DUL.toLong()
    }

    fun nextInt(bound: Int): Int = Math.floorMod(nextLong(), bound.toLong()).toInt()
}

File: fixtures/scale/dart/large/lib/util/rng.dart
/// xorshift64-ish generator; deterministic so runs are reproducible.
class Rng {
  int _state;

  Rng(int seed) : _state = seed == 0 ? 0x1E3779B9 : seed;

  int nextInt32() {
    var x = _state;
    x ^= (x << 13) & 0xFFFFFFFF;
    x ^= (x >> 7);
    x ^= (x << 17) & 0xFFFFFFFF;
    _state = x & 0xFFFFFFFF;
    return _state;
  }

  int nextInt(int bound) => nextInt32() % bound;
}
